docs/_docs/reference/experimental/typeclasses.md

```diff
@@ -0,0 +1,776 @@
+GivenSig          ::=  GivenType ['as' id] ([‘=’ Expr] | TemplateBody)
+                   |   ConstrApps ['as' id] TemplateBody
+GivenType         ::=  AnnotType {id [nl] AnnotType}
+
+TypeDef           ::=  id [TypeParamClause] TypeAndCtxBounds
+TypeParamBounds   ::=  TypeAndCtxBounds
+TypeAndCtxBounds  ::=  TypeBounds [‘:’ ContextBounds]
+ContextBounds     ::=  ContextBound | '{' ContextBound {',' ContextBound} '}'
+ContextBound      ::=  Type ['as' id]
+```
+
+
+
+## Examples
+
+
+### Example 1
+
+Here are some standard type classes, which were mostly already introduced at the start of this note, now with associated instance givens and some test code:
+
+```scala
+  // Type classes
+
+  trait Ord:
+    type Self
+    extension (x: Self)
+      def compareTo(y: Self): Int
+      def < (y: Self): Boolean = compareTo(y) < 0
+      def > (y: Self): Boolean = compareTo(y) > 0
+      def <= (y: Self): Boolean = compareTo(y) <= 0
+      def >= (y: Self): Boolean = compareTo(y) >= 0
+      def max(y: Self): Self = if x < y then y else x
+
+  trait Show:
+    type Self
+    extension (x: Self) def show: String
+
+  trait SemiGroup:
+    type Self
+    extension (x: Self) def combine(y: Self): Self
+
+  trait Monoid extends SemiGroup:
+    def unit: Self
+
+  trait Functor:
+    type Self[A] // Here, Self is a type constructor with parameter A
+    extension [A](x: Self[A]) def map[B](f: A => B): Self[B]
+
+  trait Monad extends Functor:
+    def pure[A](x: A): Self[A]
+    extension [A](x: Self[A])
+      def flatMap[B](f: A => Self[B]): Self[B]
+      def map[B](f: A => B) = x.flatMap(f `andThen` pure)
+
+  // Instances
+
+  given Int is Ord:
+    extension (x: Int)
+      def compareTo(y: Int) =
+        if x < y then -1
+        else if x > y then +1
+        else 0
+
+  given [T: Ord] => List[T] is Ord:
+    extension (xs: List[T]) def compareTo(ys: List[T]): Int =
+      (xs, ys) match
+      case (Nil, Nil) => 0
+      case (Nil, _) => -1
+      case (_, Nil) => +1
+      case (x :: xs1, y :: ys1) =>
+        val fst = x.compareTo(y)
+        if (fst != 0) fst else xs1.compareTo(ys1)
+
+  given List is Monad:
+    extension [A](xs: List[A])
+      def flatMap[B](f: A => List[B]): List[B] =
+        xs.flatMap(f)
+    def pure[A](x: A): List[A] =
+      List(x)
+
+  type Reader[Ctx] = [X] =>> Ctx => X
+
+  given [Ctx] => Reader[Ctx] is Monad:
+    extension [A](r: Ctx => A)
+      def flatMap[B](f: A => Ctx => B): Ctx => B =
+        ctx => f(r(ctx))(ctx)
+    def pure[A](x: A): Ctx => A =
+      ctx => x
+
+  // Usages
+
+  extension (xs: Seq[String])
+    def longestStrings: Seq[String] =
+      val maxLength = xs.map(_.length).max
+      xs.filter(_.length == maxLength)
+
+  extension [M[_]: Monad, A](xss: M[M[A]])
+    def flatten: M[A] =
+      xss.flatMap(identity)
+
+  def maximum[T: Ord](xs: List[T]): T =
+    xs.reduce(_ `max` _)
+
+  given [T: Ord] => T is Ord as descending:
+    extension (x: T) def compareTo(y: T) = T.compareTo(y)(x)
```

Comment:
It seems as if `descending` is not used; or perhaps the indentation of next line is wrong?

Reply:
The use if two lines below.